File: src/google/script_manager.py
import json
from googleapiclient.discovery import build
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
import os
from config.settings import GOOGLE_SCOPES, GAS_DYNAMIC_DIRECTORY

class ScriptManager:
    """
    A class to handle uploading and updating Google Apps Script projects.

    Attributes:
        creds (Credentials): The Google API credentials.
        service (Resource): The Google Apps Script API service.
    """

    SCOPES = GOOGLE_SCOPES

    def __init__(self, token=None):
        """
        Initializes the ScriptUploader with credentials and service.

        Args:
            token (str): The OAuth token to authenticate the user.
        """
        self.creds = self._authenticate(token)
        self.service = build('script', 'v1', credentials=self.creds)


    def _authenticate(self, token):
        """
        Authenticates the user using the provided token.

        Args:
            token (str): The OAuth token to authenticate the user.

        Returns:
            Credentials: The authenticated Google API credentials.
        """
        if token:
            # Create credentials from the provided token
            creds = Credentials(token)
            if not creds.valid:
                raise Exception("Invalid credentials provided.")
            return creds
        else:
            raise Exception("No token provided for authentication.")


    def update_script_content(self, script_id, generated_code=None):
        """
        Updates the script content with the provided code and all files in the 'gas' directory.

        Args:
            script_id (str): The ID of the Google Apps Script project.
            generated_code (str, optional): The JavaScript code to update the script with. Defaults to the 'get_default_dynamic_script' from settings.

        Returns:
            dict: The response from the Google Apps Script API.
        """
        # Use get_default_dynamic_script if generated_code is not provided
        if generated_code is None:
            from config.settings import get_default_dynamic_script
            generated_code = get_default_dynamic_script()

        # Start with the provided code
        files = [
            {
                'name': 'generated',
                'type': 'SERVER_JS',
                'source': generated_code
            },
            {
                'name': 'appsscript',
                'type': 'JSON',
                'source': json.dumps({
                    "timeZone": "America/New_York",
                    "exceptionLogging": "CLOUD",
                    "runtimeVersion": "V8",
                    "oauthScopes": GOOGLE_SCOPES
                })
            }
        ]

        # Add all files from the 'gas' directory and its subfolders
        gas_directory = GAS_DYNAMIC_DIRECTORY
        
        for root, _, filenames in os.walk(gas_directory):  # Use os.walk to traverse subfolders
            for filename in filenames:
                filepath = os.path.join(root, filename)
                if os.path.isfile(filepath):
                    with open(filepath, 'r') as file:
                        file_content = file.read()
                        if filename.endswith('.html'):
                            file_type = 'HTML'
                        elif filename.endswith('.js'):
                            file_type = 'SERVER_JS'
                        elif filename.endswith('.json'):
                            file_type = 'JSON'
                        else:
                            file_type = 'UNKNOWN'
                        # Use relative path for the name to maintain folder structure
                        relative_name = os.path.relpath(filepath, gas_directory)
                        files.append({
                            'name': os.path.splitext(relative_name)[0],  # Get filename without extension
                            'type': file_type,
                            'source': file_content
                        })

        # Create the request body
        request = {'files': files}

        # Execute the update request
        response = self.service.projects().updateContent(
            scriptId=script_id,
            body=request
        ).execute()
        return response

File: src/google/script-manager.js
const fs = require('fs').promises;
const path = require('path');
const { google } = require('googleapis');
const { OAuth2Client } = require('google-auth-library');
const config = require('../config');

class ScriptManager {
  
  /**
   * Generates the JavaScript code for the performAction function with the current timestamp.
   *
   * @returns {string} The JavaScript code to upload.
   */
  static getDefaultGeneratedCode() {
      const currentTime = new Date().toLocaleString('en-US', { 
        month: 'long', 
        day: 'numeric', 
        year: 'numeric', 
        hour: 'numeric', 
        minute: '2-digit', 
        hour12: true 
      }).replace(',', ' at').toLowerCase();
      
      return `
        function performAction() {
            const ui = SpreadsheetApp.getUi();
            ui.alert('This script was last generated at ${currentTime}');
        }
      `;
  }

  /**
   * Reads the Google Apps Script manifest file (appsscript.json) from the specified GAS directory.
   *
   * @param {string} gasDirectory - The directory to read the manifest file from.
   * @returns {Promise<Object>} The parsed JSON object from the manifest file.
   * @throws {Error} Throws an error if the manifest file cannot be read or parsed.
   * 
   * @static
   */
  static async readGasManifest(gasDirectory) {
    const manifestPath = path.join(gasDirectory, 'appsscript.json');
    try {
        const data = await fs.readFile(manifestPath, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        console.error("Error reading GAS manifest file:", error);
        throw new Error("Failed to read GAS manifest file.");
    }
  }

  /**
   * A class to handle uploading and updating Google Apps Script projects.
   * 
   * @param {string} token - The OAuth token to authenticate the user.
   */
  constructor(token = null) {   
    this.gasDirectory = path.join(process.cwd(), config.google.gasDirectory);
    this.creds = this._authenticate(token);
    this.service = google.script({ version: 'v1', auth: this.creds });
  }

  /**
   * Authenticates the user using the provided token.
   *
   * @param {string} token - The OAuth token to authenticate the user.
   * @returns {OAuth2Client} The authenticated Google API credentials.
   * @private
   */
  _authenticate(token) {
    if (!token) throw new Error("No token provided for authentication.");
        
    const oAuth2Client = new OAuth2Client();
    oAuth2Client.setCredentials({ access_token: token });
    if (!oAuth2Client.credentials.access_token) {
      throw new Error("Invalid credentials provided.");
    }
    
    return oAuth2Client;
  }

  /**
   * Updates the script content with the provided code and all files in the 'gas' directory.
   *
   * @param {string} scriptId - The ID of the Google Apps Script project.
   * @param {string} generatedCode - The JavaScript code to update the script with.
   * @param {string} timezone - The timezone for the script. Defaults to "America/New_York".
   * @returns {Promise<object>} The response from the Google Apps Script API.
   */
  async updateScriptContent(scriptId, generatedCode = null, timezone = "America/New_York") {
    // Use default code if generatedCode is not provided
    if (!generatedCode) {
      generatedCode = ScriptManager.getDefaultGeneratedCode();
    }
    
    const manifest = await ScriptManager.readGasManifest(this.gasDirectory);
    manifest.timeZone = timezone;

    // Start with the provided code
    const files = [
      {
        name: 'generated',
        type: 'SERVER_JS',
        source: generatedCode
      },
      {
        name: 'appsscript',
        type: 'JSON',
        source: JSON.stringify(manifest)
      }
    ];

    try {
      // Add all files from the 'gas' directory and its subfolders
      await this.addFilesFromDirectory(this.gasDirectory, files);
      
      // Create the request body
      const request = { files };

      // Execute the update request
      const response = await this.service.projects.updateContent({
        scriptId,
        requestBody: request
      });
      
      return response.data;
    } catch (error) {
      console.error("Error updating script content:", error);
      return { error: error.message };
    }
  }

  /**
   * Recursively adds files from a directory to the files array.
   *
   * @param {string} directory - The directory to read files from.
   * @param {Array} files - The array to add files to.
   * @returns {Promise<void>}
   * @private
   */
  async addFilesFromDirectory(directory, files) {
    try {
      const entries = await fs.readdir(directory, { withFileTypes: true });
      
      for (const entry of entries) {
        const fullPath = path.join(directory, entry.name);
        
        if (entry.isDirectory()) {
          // Recursively process subdirectories
          await this.addFilesFromDirectory(fullPath, files);
        } else {
          // Process files
          const fileContent = await fs.readFile(fullPath, 'utf8');
          let fileType;
          
          if (entry.name.endsWith('.html')) {
            fileType = 'HTML';
          } else if (entry.name.endsWith('.js')) {
            fileType = 'SERVER_JS';
          } else {
            continue;
          }
          
          // Use relative path for the name to maintain folder structure
          const relativePath = path.relative(this.gasDirectory, fullPath);
          const nameWithoutExt = path.join(
            path.dirname(relativePath),
            path.basename(relativePath, path.extname(relativePath))
          ).replace(/\\/g, '/');
          
          files.push({
            name: nameWithoutExt,
            type: fileType,
            source: fileContent
          });
        }
      }
    } catch (error) {
      console.error("Error reading directory:", error);
      throw error;
    }
  }
}

module.exports = ScriptManager;
